lib/libvfio-user.c

```diff
@@ -286,55 +287,94 @@ is_valid_region_access(vfu_ctx_t *vfu_ctx, size_t size, uint16_t cmd,
 }
 
 static int
-handle_region_access(vfu_ctx_t *vfu_ctx, uint32_t size, uint16_t cmd,
-                     void **data, size_t *len,
-                     struct vfio_user_region_access *ra)
+handle_region_access(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
 {
+    struct vfio_user_region_access *in_ra = msg->in_data;
+    struct vfio_user_region_access *out_ra;
     ssize_t ret;
     char *buf;
 
     assert(vfu_ctx != NULL);
-    assert(data != NULL);
-    assert(ra != NULL);
+    assert(msg != NULL);
 
-    if (!is_valid_region_access(vfu_ctx, size, cmd, ra)) {
+    if (!is_valid_region_access(vfu_ctx, msg->in_size, msg->hdr.cmd, in_ra)) {
         return ERROR_INT(EINVAL);
     }
 
-    if (ra->count == 0) {
+    if (in_ra->count == 0) {
         return 0;
     }
 
-    *len = sizeof(*ra);
-    if (cmd == VFIO_USER_REGION_READ) {
-        *len += ra->count;
+    msg->out_size = sizeof(*in_ra);
+    if (msg->hdr.cmd == VFIO_USER_REGION_READ) {
+        msg->out_size += in_ra->count;
     }
-    *data = calloc(1, *len);
-    if (*data == NULL) {
+    msg->out_data = calloc(1, msg->out_size);
+    if (msg->out_data == NULL) {
         return -1;
     }
-    if (cmd == VFIO_USER_REGION_READ) {
-        buf = (char *)(((struct vfio_user_region_access*)(*data)) + 1);
+
+    out_ra = msg->out_data;
+    out_ra->region = in_ra->region;
+    out_ra->offset = in_ra->offset;
+    out_ra->count = in_ra->count;
+
+    if (msg->hdr.cmd == VFIO_USER_REGION_READ) {
+        buf = (char *)(&out_ra->data);
     } else {
-        buf = (char *)(ra + 1);
+        buf = (char *)(&in_ra->data);
     }
 
-    ret = region_access(vfu_ctx, ra->region, buf, ra->count, ra->offset,
-                        cmd == VFIO_USER_REGION_WRITE);
+    ret = region_access(vfu_ctx, in_ra->region, buf, in_ra->count,
+                        in_ra->offset, msg->hdr.cmd == VFIO_USER_REGION_WRITE);
 
-    if (ret != ra->count) {
+    if (ret != in_ra->count) {
         vfu_log(vfu_ctx, LOG_ERR, "failed to %s %#x-%#lx: %s",
```

Comment:
There was a bug here: we'd leak *data (found via python tests). Shouldn't occur now we use ->out_data.